doc/development.rst

```diff
@@ -292,6 +292,72 @@ Here the `barebox` state simply cycles the board and activates the driver, while
 the `shell` state uses the barebox state to cycle the board and than boot the
 linux kernel. The `off` states switch the power off.
 
+
+Tips for Writing and Debugging Tests
+------------------------------------
+
+Live-Reading Console Output
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When starting labgrid with ``--lg-log`` option, it will dump the input from the
+serial driver to a file in specified directory::
+
+  $ pytest .. --lg-log=logdir test-dir/
+
+This can help understanding what happened and why it happened.
+However, when debugging tests, it might be more helpful to get a live
+impression of what is going on.
+For this, you can use ``tail -f`` to read the content written to the log file
+as if you would be connected to the devices serial console (except that it is
+read-only)::
+
+  $ tail -f logdir/console_file
+
+For getting information about timing, the ``annotate-output`` command turned
+out to be quite helpful. On Debian it comes with the ``devscripts`` package and
+you can install it with::
+
+  $ apt-get install devscripts
+
+To use it, run::
+
+  $ annotate-output tail -f logdir/console_file
+
+This will print your system time before each line, allowing you to both see
+relative delays between steps in your tests as well as absolute timing
+information to conclude about correlation of things happening in your test
+environment.
+
+Dealing With Kernel Log Verbosity
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For testing your Linux system it can be quite annoying if the kernel outputs
+verbosely to the console you use for testing.
+Note that a too verbose kernel can break tests as kernel logs will pollute the
+expected command outputs making it unreadable for labgrid regexes.
+
+However, as the shell drivers and most of the tests will depend on seeing
+console output of what is going on, we cannot simply turn off kernel logging.
```

Comment:
Hm, we are doing exactly this once we are logged into the console. Should we change this behaviour?

Reply:
There is still a small race condition if we get kernel output before disabling printk to the console. This should be reflected in the docs.

Reply:
I've added a note box trying to reflect this.